### ISSUE TITLE
Approximation leads to invalid double with mpq_class (at least)

### ISSUE THREAD

**Issue by cedricpradalier** [CGAL/cgal#3507]
I used Nef_S2 with mpq_class FT to recursively build polygon on the unit sphere. The to_double conversion tends to lead to extremely small numbers (1e-50, 1e-120... ) with means that the normalization is not reliable. 

I've modified convert as follows:
```c++
    typename R::FT xa = CGAL::abs(v.x());
    typename R::FT ya = CGAL::abs(v.y());
    typename R::FT za = CGAL::abs(v.z());
    typename R::FT m = std::max(std::max(xa,ya),za);
    if (m==0) {
        return VVector(0,0,0);
    } else {
        double xd = CGAL::to_double(v.x()/m);
        double yd = CGAL::to_double(v.y()/m);
        double zd = CGAL::to_double(v.z()/m);
        VVector u(xd,yd,zd);
        return u / CGAL_NTS sqrt(u*u) ; // normalize
    }
```

With Gmpq or mpq_class this garantees that the numbers are not too far from 1 before conversion to double, without loss of precision. Otherwise, the conversion to degenerate triangles leads to infinite loops in refine. As a backup, I've added a recursive depth counter to refine, and the following lines:
```c++
    if ((depth>10)||std::isnan(angle[0])||std::isnan(angle[1])||std::isnan(angle[2])) {
        // Degenerate case. Should not be there, but this means a 
        // degenerate triangle slipped through
        return;
    }
```
I dont think they are necessary with the conversion above, but it at least guarantees that the code returns. 

https://github.com/CGAL/cgal/blob/84151fe9b3614b0e6387ac8316d47fc92809ab6b/Nef_S2/include/CGAL/Nef_S2/Sphere_geometry_OGL.h#L82

**Comment by sloriot** [comment]
Would 7c6450192e fixes the issue? I can't update the convert function has the behavior would no longer be the same (because of the normalization).

**Comment by lrineau** [comment]
Fixed by #3510.
